x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go

```diff
@@ -0,0 +1,200 @@
+//  http.request.body.bytes
+//  http.request.mime_type
+//  event.original (the full request and body from httputil.DumpRequestOut)
+//
+// Fields logged in responses:
+//  http.response.status_code
+//  http.response.body.content
+//  http.response.body.bytes
+//  http.response.mime_type
+//  event.original (the full response and body from httputil.DumpResponse)
+//
+func (rt *LoggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Create a child logger for this request.
+	log := rt.logger.With(
+		zap.String("transaction.id", rt.nextTxID()),
+	)
+
+	if v := req.Context().Value(TraceIDKey); v != nil {
+		if traceID, ok := v.(string); ok {
+			log = log.With(zap.String("trace.id", traceID))
+		}
+	}
+
+	reqParts := []zapcore.Field{
+		zap.String("url.original", req.URL.String()),
+		zap.String("url.scheme", req.URL.Scheme),
+		zap.String("url.path", req.URL.Path),
+		zap.String("url.domain", req.URL.Hostname()),
+		zap.String("url.port", req.URL.Port()),
+		zap.String("url.query", req.URL.RawQuery),
+		zap.String("http.request.method", req.Method),
+		zap.String("user_agent.original", req.Header.Get("User-Agent")),
+	}
+	var (
+		body   []byte
+		err    error
+		errors []error
+	)
+	req.Body, body, err = copyBody(req.Body)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to read request body: %w", err))
+	} else {
+		reqParts = append(reqParts,
+			zap.ByteString("http.request.body.content", body),
+			zap.Int("http.request.body.bytes", len(body)),
+			zap.String("http.request.mime_type", req.Header.Get("Content-Type")),
+		)
+	}
+	message, err := httputil.DumpRequestOut(req, true)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to dump request: %w", err))
+	} else {
+		reqParts = append(reqParts, zap.ByteString("event.original", message))
+	}
+	if errors != nil {
+		reqParts = append(reqParts, zap.Errors("error.message", errors))
+	}
+	log.Debug("HTTP request", reqParts...)
+
+	resp, err := rt.transport.RoundTrip(req)
+	if err != nil {
+		log.Debug("HTTP response error", zap.NamedError("error.message", err))
+		return resp, err
+	}
+	if resp == nil {
+		log.Debug("HTTP response error", noResponse)
+		return resp, err
+	}
+	respParts := append(reqParts[:0],
+		zap.Int("http.response.status_code", resp.StatusCode),
+	)
+	errors = errors[:0]
+	resp.Body, body, err = copyBody(resp.Body)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to read response body: %w", err))
+	} else {
+		respParts = append(respParts,
+			zap.ByteString("http.response.body.content", body),
+			zap.Int("http.response.body.bytes", len(body)),
+			zap.String("http.response.mime_type", resp.Header.Get("Content-Type")),
+		)
+	}
+	message, err = httputil.DumpResponse(resp, true)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to dump response: %w", err))
+	} else {
+		respParts = append(respParts, zap.ByteString("event.original", message))
+	}
+	if errors != nil {
```

Comment:
I think this needs to check for `len(errors) > 0`. In the case where the request produces an error and then we zero out the length this will be non-nil. In general I don't like to make a distinction between nil and empty slices so I tend to always check for length.

Reply:
Good point. I wrote the first one knowing that the zero length would be nil and then didn't reassess when I duplicated it.